src/NonceTracker.ts

```diff
@@ -89,25 +88,42 @@ export interface Transaction {
 }
 
 export class NonceTracker {
-  private provider: Record<string, unknown>;
+  #provider: Record<string, unknown>;
 
-  private blockTracker: PollingBlockTracker;
+  #blockTracker: PollingBlockTracker;
 
-  private web3: typeof Web3Provider;
+  readonly #getPendingTransactions: (address: string) => Transaction[];
 
-  private getPendingTransactions: (address: string) => Transaction[];
+  readonly #getConfirmedTransactions: (address: string) => Transaction[];
 
-  private getConfirmedTransactions: (address: string) => Transaction[];
-
-  private lockMap: Record<string, Mutex>;
+  readonly #lockMap: Record<string, Mutex>;
 
   constructor(opts: NonceTrackerOptions) {
-    this.provider = opts.provider;
-    this.blockTracker = opts.blockTracker;
-    this.web3 = new Web3Provider(opts.provider);
-    this.getPendingTransactions = opts.getPendingTransactions;
-    this.getConfirmedTransactions = opts.getConfirmedTransactions;
-    this.lockMap = {};
+    this.#provider = opts.provider;
+    this.#blockTracker = opts.blockTracker;
+    this.#getPendingTransactions = opts.getPendingTransactions;
+    this.#getConfirmedTransactions = opts.getConfirmedTransactions;
+    this.#lockMap = {};
+  }
+
+  /**
+   * Allows changing the provider and block tracker after insrtantiation. As the blockTracker also has a provider, they are updated atomically.
+   *
+   * @param opts - new props
+   * @param opts.provider - An ethereum provider
+   * @param opts.blockTracker - An instance of @metamask/eth-block-tracker
+   */
+  setProvider({
```

Comment:
I am very easily to convince to remove the `setProvider` function alltogether, if there is no current use for it. Put it in to make sure there are no compatibility concerns by default.